### PR TITLE
Add api schema to schema.

### PR DESCRIPTION
**Pull request by BrynCooke**
Creating the api schema is done on parse, and blockingly calls out to deno. It should be OK because we load new schemas in the state machine and won't actually block the main request handlers.

Query shaping now operates against the api schema.

Introspection is performed against API schema on the Node side.

Fixed all the broken tests as we now require valid graphql schemas everywhere.

Related PR https://github.com/apollographql/federation-rs/pull/47

Fixes #573

### CODE REVIEW

File: apollo-router-core/src/spec/schema.rs [hunk 1]
```diff
@@ -12,287 +14,318 @@ pub struct Schema {
...
+            let document = tree.document();
+            let mut subtype_map: HashMap<String, HashSet<String>> = Default::default();
+            let mut subgraphs = HashMap::new();
+
+            // the logic of this algorithm is inspired from the npm package graphql:
+            // https://github.com/graphql/graphql-js/blob/ac8f0c6b484a0d5dca2dc13c387247f96772580a/src/type/schema.ts#L302-L327
+            // https://github.com/graphql/graphql-js/blob/ac8f0c6b484a0d5dca2dc13c387247f96772580a/src/type/schema.ts#L294-L300
+            // https://github.com/graphql/graphql-js/blob/ac8f0c6b484a0d5dca2dc13c387247f96772580a/src/type/schema.ts#L215-L263
+            for definition in document.definitions() {
+                macro_rules! implements_interfaces {
```
Comment (bnjjj) [inline apollo-router-core/src/spec/schema.rs:66]: Am I the only one feeling bad to see a macro_rules at this place ? I find this harder to read

Comment (BrynCooke) [reply apollo-router-core/src/spec/schema.rs:66]: Agree, but it's out of the remit of this PR to change. It's just indenting that got modified here.

File: apollo-router-core/src/error.rs [hunk 2]
```diff
@@ -270,6 +270,8 @@ pub enum SchemaError {
     IoError(#[from] std::io::Error),
     /// Parsing error(s).
     Parse(ParseErrors),
+    /// Api error(s): {0}
+    Api(String),
```
Comment (o0Ignition0o) [inline apollo-router-core/src/error.rs:274]: What does this mean ? ^^'

File: apollo-router/src/state_machine.rs [hunk 3]
```diff
@@ -548,14 +554,16 @@ mod tests {
...
+                                            routing_url: Url::parse("http://accounts/graphql")
+                                                .unwrap(),
                                             layers: Vec::new(),
                                         }
                                     ),
                                     (
                                         "products".to_string(),
                                         Subgraph {
-                                            routing_url: Url::parse("http://accounts/graphql").unwrap(),
+                                            routing_url: Url::parse("http://accounts/graphql")
```
Comment (o0Ignition0o) [inline apollo-router/src/state_machine.rs:565]: thought this was incorrect (I was expecting `http://products/graphql`) but it's not even checked. not sure `extract_routing_urls` isn't a misleading title, or if this is being checked somewhere. do you want me to investigate?

File: apollo-router-core/src/error.rs (hunk 2 above)
Comment (o0Ignition0o) [reply apollo-router-core/src/error.rs:274]: ok now i've seen the use, is this a "ConvertToApiSchemaError" ?

**Review by o0Ignition0o — APPROVED** [review]

File: apollo-router/src/state_machine.rs (hunk 3 above)
Comment (BrynCooke) [reply apollo-router/src/state_machine.rs:565]: I think this will disappear soon as routing URL is being replaced with a plugin.

Comment (bnjjj) [reply apollo-router/src/state_machine.rs:565]: yes

**Pull request merged**